File: day1/day1.exs
defmodule Advent do
  def add_sum_to_list(n, []), do: [n]
  def add_sum_to_list(n, acc) do
    acc ++ [n + (List.last(acc))]
  end

  def find_repeat(sequence) do
    find_repeat(sequence, 9999999, 0)
  end

  def find_repeat([n | seq], place, repeat) do
    {new_place, new_repeat} = check_column(n, seq, place, repeat, 0)
    if new_place < place do
      find_repeat(seq, new_place, new_repeat)
    else
      find_repeat(seq, place, repeat)
    end
  end

  def find_repeat([], _, repeat) do
    repeat
  end

  def check_column(_n1, [], place, repeat, _index), do: {place, repeat}
  def check_column(_n1, [_n2], place, repeat, _index), do: {place, repeat}

  def check_column(n1, [n2 | seq], place, repeat, index) do
    sum = List.last(seq)
    diff = abs(n1 - n2)
    case rem(diff, sum) do
      0 ->
        new_place = diff/sum + index
        if new_place < place do
          new_repeat = if n1 > n2, do: n1, else: n2
          check_column(n1, seq, new_place, new_repeat, index+1)
        else
          check_column(n1, seq, place, repeat, index+1)
        end
      _ ->
        check_column(n1, seq, place, repeat, index+1)
    end
  end
end

# content = "+3, +3, +4, -2, -4"
# sums = String.split(content, ", ")
{:ok, content} = File.read("day1_input.txt")
sums = String.split(content, "\n")
|> Enum.map(fn n -> String.to_integer(n) end)
|> Enum.reduce([], fn n, acc -> Advent.add_sum_to_list(n, acc) end)

IO.puts List.last(sums)

repeat = Advent.find_repeat(sums)
IO.puts repeat

File: day1/day1.rb
# There 1,040 elements on the list
# Sum is 569
# Average is 0.54
# Largest number is 76,787
# Smallest number -77371

# So for each entry in the input, we can make a column
# And the output for that column is = (cycle number) * (total sum) + (column's sum in the first cycle)
# This is a linear equation - ax + b. And every column has the same a - the sum of the entire sequence
# I'm not looking for intersections, which is good because there are none. All these lines are parallel
# I'm looking for whole number values where these two entries produce the same value

# Is it possible to have a combination of values which will never produce the same values?
# Yes: 
# 2x + 1: 1 3 5 7 9 11
# 2x + 0: 0 2 4 6 8 10
# Those two sequences will never share a number

# What are the attributes of sequences which will produce the same values?
# 4x + 5:  *5* 9 13  17
# 4x - 3:  -3  1 *5*  9
# I think the important attribute is that the difference between these two values is divisible by the sum, (5 - -3 = 8, 8 % 4 = 0).

# Given two sequences that I think will share a value, how to determine what that value is?
# In the examples that I've tried, the first value always showed up in the first cycle, the larger value
# So let's assume that's true, so that makes it:
# a * x + b1 = b2
# x = (b2 - b1)/a
# x = round we're on

require 'benchmark'

time = Benchmark.measure do
  input = File.open("day1_input.txt").map {|n| n.to_i}

  sums = input.reduce([]) {|s, n| s = s + [n + (s.last || 0)]}
  sum = sums.last

  p "Sum: #{sum}"

  repeat = 0
  round = 9999999
  place = sums.length + 1
  sums.each_with_index do |i, index|
    sums.slice(index+1, sums.length).each do |j|
      diff = (i - j).abs
      if diff % sum == 0
        match_round = diff/sum
        if match_round < round || (match_round == round && index < place)
          repeat = i > j ? i : j
          round = match_round
          place = index
        end
      end
    end
  end
  # Cases this algorithm won't catch:
    # Repeating value is 0 - it'll find the next value
    # If the sum is negative

  p "Repeating Frequency: #{repeat}"
end

p ""
p " cpu        system     total    (real)"
puts time
